src/main/scala/org/scalacheck/Gen.scala

```diff
@@ -763,115 +763,78 @@ object Gen extends GenArities{
   )
 
   lazy val calendar: Gen[java.util.Calendar] = {
-    import java.util.{Calendar, Date}
-
-    val MinYearForCalendar = {
-      val c = Calendar.getInstance()
-      c.getGreatestMinimum(Calendar.YEAR)
-    }
-
-    val MaxYearForCalendar = {
-      val c = Calendar.getInstance()
-      c.getLeastMaximum(Calendar.YEAR)
-    }
-
-    def buildLastDayOfMonth(c: Calendar): Calendar = {
-      val lastDayOfMonth = c.getActualMaximum(Calendar.DAY_OF_MONTH)
-      c.set(Calendar.DAY_OF_MONTH, lastDayOfMonth)
-      c
-    }
-
-    def buildFirstDayOfMonth(c: Calendar): Calendar = {
-      c.set(Calendar.DAY_OF_MONTH, 1)
-      c
-    }
-
-    def buildFirstDayOfYear(year: Int): Calendar = {
-      val c = Calendar.getInstance()
-      c.set(year, 0, 1)
-      c
-    }
-
-    def buildLastDayOfYear(year: Int): Calendar = {
-      val c = Calendar.getInstance()
-      c.set(year, 11, 31)
-      c
-    }
-
-    def buildNearestLeapDate(year: Int): Calendar = {
-      val c = Calendar.getInstance()
-      c.set(closestLeapYear(year), 1, 29)
-      c
-    }
-
-    def closestLeapYear(year: Int): Int = {
-      var currentYear = year match {
-        case y if (y + 4) > MaxYearForCalendar => MaxYearForCalendar - 5
-        case _ => year
+    import java.util.{ Calendar, Date, TimeZone, Locale }
+    import Calendar._
+
+    def adjust(c: Calendar)(f: Calendar => Unit): Calendar = { f(c); c }
+
+    // We want to be sure we always initialize the calendar's time. By
+    // default, Calendar.getInstance uses the system time. We always
+    // overwrite it with a determinisitcally-geneated time to be sure
+    // that calendar generation is also deterministic.
+    //
+    // We limit the time (in milliseconds) because extreme values will
+    // cause Calendar.getTime calls to fail. This range is relatively
+    // large but safe:
+    //
+    //   -62135751600000 is 1 CE
+    //    64087186649116 is 4000 CE
+    val calendar: Gen[Calendar] =
+      Gen.chooseNum(-62135751600000L, 64087186649116L).map { t =>
+        adjust(Calendar.getInstance)(_.setTimeInMillis(t))
       }
-      while (!isLeapYear(currentYear)) {
-        currentYear += 1
-      }
-      currentYear
-    }
-
-    def isLeapYear(year: Int): Boolean = {
-      val cal = Calendar.getInstance()
-      cal.set(Calendar.YEAR, year)
-      cal.getActualMaximum(Calendar.DAY_OF_YEAR) > 365
-    }
-
-    val yearGen = Gen.chooseNum(MinYearForCalendar, MaxYearForCalendar)
-
-    val basicCalendarGen: Gen[Calendar] = for {
-      l <- Gen.chooseNum(Long.MinValue, Long.MaxValue)
-      now = new Date
-      d = new Date(now.getTime + l)
-      c = Calendar.getInstance()
-    } yield {
-      c.setTimeInMillis(d.getTime)
-      c
-    }
 
-    val calendarBeginningOfDayGen: Gen[Calendar] = for {
```

Comment:
btw @tixxit all these generators make the `Gen` / `Arbitrary` distinction much clearer.